python/ql/src/semmle/python/frameworks/Stdlib.qll

```diff
@@ -864,6 +864,230 @@ private module Stdlib {
+  private API::Node pathlib() { result = API::moduleImport("pathlib") }
+
+  /**
+   * Gets a name of a constructor for a `pathlib.Path` object.
+   * We include the pure paths, as they can be "exported" (say with `as_posix`) and then used to acces the underlying file system.
+   */
+  private string pathlibPathConstructor() {
+    result in ["Path", "PurePath", "PurePosixPath", "PureWindowsPath", "PosixPath", "WindowsPath"]
+  }
+
+  /**
+   * Gets a name of an attribute of a `pathlib.Path` object that is also a `pathlib.Path` object.
+   */
+  private string pathlibPathAttribute() { result in ["parent"] }
+
+  /**
+   * Gets a name of a method of a `pathlib.Path` object that returns a `pathlib.Path` object.
+   */
+  private string pathlibPathMethod() {
+    result in ["absolute", "relative_to", "rename", "replace", "resolve"]
+  }
+
+  /**
+   * Gets a name of a method of a `pathlib.Path` object that modifies a `pathlib.Path` object based on new data.
+   */
+  private string pathlibPathInjection() {
+    result in ["joinpath", "with_name", "with_stem", "with_suffix"]
+  }
+
+  /**
+   * Gets a name of an attribute of a `pathlib.Path` object that exports information about the `pathlib.Path` object.
+   */
+  private string pathlibPathAttributeExport() {
+    result in ["drive", "root", "anchor", "name", "suffix", "stem"]
+  }
+
+  /**
+   * Gets a name of a method of a `pathlib.Path` object that exports information about the `pathlib.Path` object.
+   */
+  private string pathlibPathMethodExport() { result in ["as_posix", "as_uri"] }
+
+  /**
+   * Gets a reference to a `pathlib.Path` object.
+   * This type tracker makes the monomorphic API use assumption.
+   */
+  private DataFlow::LocalSourceNode pathlibPath(DataFlow::TypeTracker t) {
+    // Type construction
+    t.start() and
+    result = pathlib().getMember(pathlibPathConstructor()).getACall()
+    or
+    // Type-preserving call
+    exists(DataFlow::AttrRead returnsPath, DataFlow::TypeTracker t2 |
+      returnsPath.getAttributeName() = pathlibPathMethod() and
+      returnsPath.getObject().getALocalSource() = pathlibPath(t2) and
+      t2.end()
+    |
+      t.start() and
+      result.(DataFlow::CallCfgNode).getFunction() = returnsPath
+    )
+    or
+    // Type-preserving attribute
+    exists(DataFlow::AttrRead isPath, DataFlow::TypeTracker t2 |
+      isPath.getAttributeName() = pathlibPathAttribute() and
+      isPath.getObject().getALocalSource() = pathlibPath(t2) and
+      t2.end()
+    |
+      t.start() and
+      result = isPath
+    )
+    or
+    // Data injection
+    exists(
+      BinaryExprNode slash, DataFlow::Node right, DataFlow::Node left, DataFlow::TypeTracker t2
+    |
+      slash.getOp() instanceof Div and
+      right.asCfgNode() = slash.getRight() and
+      left.asCfgNode() = slash.getLeft() and
+      (
+        left.getALocalSource() = pathlibPath(t2)
+        or
+        right.getALocalSource() = pathlibPath(t2)
```

Comment:
This can be expressed more easily using `BinaryExprNode::getAnOperand`.

Reply:
[](url)
```suggestion
      right.asCfgNode() = slash.getRight() and
      left.asCfgNode() = slash.getLeft() and
      (
        left.getALocalSource() = pathlibPath(t2)
        or
        right.getALocalSource() = pathlibPath(t2)
```

Reply:
Indeed, thanks.